apps/channels/audio.py

```diff
@@ -17,6 +20,60 @@ def get_transcript(audio: BytesIO) -> str:
 
 
 def synthesize_voice(text: str, synthetic_voice: SyntheticVoice) -> Tuple[BytesIO, float]:
+    if synthetic_voice.service == "AWS":
+        return AWS_synthesize_voice(text, synthetic_voice)
+    elif synthetic_voice.service == "azure":
+        return azure_synthesize_voice(text, synthetic_voice)
+
+    raise AudioSynthesizeException(f"Unrecognized audio service: {synthetic_voice.service}")
+
+
+def azure_synthesize_voice(text: str, synthetic_voice: SyntheticVoice) -> Tuple[BytesIO, float]:
+    """
+    Calls Azure's cognitive speech services to convert the text to speech using the synthetic_voice
+    """
+    try:
+        speech_config = speechsdk.SpeechConfig(
+            subscription=settings.AZURE_SUBSCRIPTION_KEY, region=settings.AZURE_REGION
+        )
+
+        temp_file = tempfile.NamedTemporaryFile(delete=False, suffix=".wav")
```

Comment:
why the `delete=False`? my understanding from looking at the code is that speechsdk writes to the temp_file, but if `delete=True`, that would remove the temp_file again? Is that what's happening? :thinking:

Reply:
You can try setting delete=True (the default) and see if it's still working. If it is, we can chuck the `os.unlink(temp_file.name)` pieces

Reply:
According to my coding partner, if you have `delete=True`, then the file is deleted as soon as it is closed. Since I'm re-opening the file to read out the contents and pass it to `pydub`, then I need to not have it automatically deleted. I didn't completely understand. Looking at the [docs](https://docs.python.org/3/library/tempfile.html) now:

> If delete is true (the default) and delete_on_close is true (the default), the file is deleted as soon as it is closed. If delete is true and delete_on_close is false, the file is deleted on context manager exit only, or else when the [file-like object](https://docs.python.org/3/glossary.html#term-file-like-object) is finalized. Deletion is not always guaranteed in this case (see [object.__del__()](https://docs.python.org/3/reference/datamodel.html#object.__del__)). If delete is false, the value of delete_on_close is ignored.

I don't really understand the context manager, so defer to you all about how to interpret this. My coding partner thinks an alternative approach (which might be cleaner and remove all of the clean up code) would be:

```
with tempfile.NamedTemporaryFile(delete=True, delete_on_close=False) as tmp:
    # Work with the temp file
    pass
# Temp file is deleted here when the with block is exited
```

I assume that an exception generated within a `with` block would also trigger the `__exit__()` and clean up the temp file?

Reply:
> I assume that an exception generated within a with block would also trigger the __exit__() and clean up the temp file?

yes. if possible, I like the idea to work with everything inside the context manager. Another approach is to use BytesIO?